app/models/url.rb

```diff
@@ -0,0 +1,21 @@
+require "uri"
+class Url < ActiveRecord::Base
+  validates :long, :uniqueness => true, format: { with: /\A#{URI::regexp}\z/}
+  before_save :convert
+
+  private
+
+  def convert
+     range = [*'0'..'9', *'a'..'z', *'A'..'Z']
+     unless self.short
+        short_url = Array.new(8){range.sample}.join
```

Comment:
Also, you should be sure to have the database guarantee the uniqueness of `short_url`. This doesn't quite do it -- another server might have written a row with the same `short_url` in between when you try to `Url.find_by(short_url:...)` and when your url actually hits the database.

Ideally, you'd probably want the database to generate unique short forms itself, but that requires custom SQL (I think) and is a bit out of scope. [Here's](http://blog.00null.net/easily-generating-random-strings-in-postgresql/) a discussion of how you'd do it.